### DISCUSSION TITLE
Add Sapphire Elastika to Cardinal?

### DISCUSSION THREAD

**Discussion by cosinekitty** [DISTRHO/Cardinal#450] [General]
I would like to submit my [Sapphire](https://github.com/cosinekitty/sapphire) plugin, particularly its module [Elastika](https://github.com/cosinekitty/sapphire/blob/main/Elastika.md), as a candidate for inclusion in Cardinal. Elastika was recently featured in a [video by Omri Cohen](https://www.youtube.com/watch?v=DEoMeNHOVgE), and with the resulting attention, people are starting to ask me if it can be added to Cardinal. So here I am.

Some quick facts:

* Sapphire is licensed GPLv3+.
* It uses a few very simple knob and slider graphics that would have to be replaced for licensing reasons. Hopefully this is minor.
* All other artwork in the panel design falls under GPLv3+.

I will be happy to do as much of the porting work as would be helpful. I already have Cardinal building and running on my Linux system, and I'm reading the developer documentation now. I just wanted to post this issue to find out if a pull request would be welcomed, or is there another route to including a new module in Cardinal?

Thank you!

**Comment by falkTX** [comment]
Add it to the list https://github.com/DISTRHO/Cardinal/wiki/Possible-modules-to-include

The issue tracker is not meant to be used for requesting modules.

**Comment by falkTX** [comment]
regarding inclusion, we can discuss it of course, I just dont want the issue tracker to be full of module requests.
see https://github.com/DISTRHO/Cardinal/discussions/63 for an overview of the process for adding modules

**Comment by falkTX** [comment]
PS: converted ticket to discussion

**Comment by cosinekitty** [comment]
My apologies for the noise in the issue tracker. I went ahead and created PR #451. I noticed CPU overhead of Elastika is much higher than I encounter in VCV Rack. Elastika does some heavy duty SIMD number crunching. Is there something special I need to do to improve performance in that area?

**Comment by falkTX** [comment]
@cosinekitty there is a memory related issue detected by CI

```
==146132== Conditional jump or move depends on uninitialised value(s)
==146132==    at 0x1B806C1: setAgcEnabled (elastika_engine.hpp:354)
==146132==    by 0x1B806C1: Sapphire::ElastikaEngine::initialize() (elastika_engine.hpp:278)
==146132==    by 0x1B82029: ElastikaEngine (elastika_engine.hpp:245)
==146132==    by 0x1B82029: ElastikaModule::ElastikaModule() (elastika.cpp:76)
==146132==    by 0x1B83BBF: rack::CardinalPluginModel<ElastikaModule, ElastikaWidget>::createModule() (helpers.hpp:63)
==146132==    by 0x9516BF: CardinalDISTRHO::CardinalUI::uiIdle() (CardinalUI.cpp:484)
==146132==    by 0x961241: plugin_idle (DistrhoUIInternal.hpp:248)
==146132==    by 0x961241: CardinalDISTRHO::runSelfTests() (DistrhoPluginJACK.cpp:886)
==146132==    by 0x82AAED: main (DistrhoPluginJACK.cpp:993)
==146132==  Uninitialised value was created by a heap allocation
==146132==    at 0x6AFFE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==146132==    by 0x1B83BB4: rack::CardinalPluginModel<ElastikaModule, ElastikaWidget>::createModule() (helpers.hpp:63)
==146132==    by 0x9516BF: CardinalDISTRHO::CardinalUI::uiIdle() (CardinalUI.cpp:484)
==146132==    by 0x961241: plugin_idle (DistrhoUIInternal.hpp:248)
==146132==    by 0x961241: CardinalDISTRHO::runSelfTests() (DistrhoPluginJACK.cpp:886)
==146132==    by 0x82AAED: main (DistrhoPluginJACK.cpp:993)
```

please fix and create a PR updating the commit hash of your Sapphire modules. thanks!

**Reply by cosinekitty** [reply]
Very nice catch! I should use Valgrind like this also. I will fix the issue.

**Reply by cosinekitty** [reply]
Submitted in #452.

**Comment by falkTX** [comment]
your 2 modules are now in https://cardinal.kx.studio/ too, just like magic :mage:

**Reply by cosinekitty** [reply]
Awesome! Thank you! I didn't know Cardinal could run in a browser. Really cool. That must be C++ compiled to WebAssembly?

**Reply by falkTX** [reply]
yes, emscripten does the magic, plus of course some patching in Rack to allow such things but it is not much

**Reply by Astrinvega** [reply]
Cardinal running in a browser....amazing
